llama-index-integrations/llms/llama-index-llms-mistralai/llama_index/llms/mistralai/base.py

```diff
@@ -252,6 +266,86 @@ def stream_complete(
         stream_complete_fn = stream_chat_to_completion_decorator(self.stream_chat)
         return stream_complete_fn(prompt, **kwargs)
 
+    def _get_tool_call(
+        self,
+        response: ChatResponse,
+    ) -> ToolCall:
+        tool_calls = response.message.additional_kwargs.get("tool_calls", [])
+
+        if len(tool_calls) < 1:
+            raise ValueError(
+                f"Expected at least one tool call, but got {len(tool_calls)} tool calls."
+            )
+
+        # TODO: support more than one tool call?
+        tool_call = tool_calls[0]
+        if not isinstance(tool_call, ToolCall):
+            raise ValueError("Invalid tool_call object")
+
+        if tool_call.type != "function":
+            raise ValueError("Invalid tool type. Unsupported by Mistralai.")
+
+        return tool_call
+
+    def _call_tool(
+        self,
+        tool_call: ToolCall,
+        tools_by_name: Dict[str, "BaseTool"],
+        verbose: bool = False,
+    ) -> "AgentChatResponse":
+        from llama_index.core.chat_engine.types import AgentChatResponse
+        from llama_index.core.tools.calling import call_tool
+
+        arguments_str = tool_call.function.arguments
+        name = tool_call.function.name
+        if verbose:
+            print("=== Calling Function ===")
+            print(f"Calling function: {name} with args: {arguments_str}")
+        tool = tools_by_name[name]
+        argument_dict = json.loads(arguments_str)
+
+        tool_output = call_tool(tool, argument_dict)
+
+        return AgentChatResponse(response=tool_output.content, sources=[tool_output])
+
+    def predict_and_call(
+        self,
+        tools: List["BaseTool"],
+        user_msg: Optional[Union[str, ChatMessage]] = None,
+        chat_history: Optional[List[ChatMessage]] = None,
+        verbose: bool = False,
+        **kwargs: Any,
+    ) -> "AgentChatResponse":
+        if not self.metadata.is_function_calling_model:
+            return super().predict_and_call(
```

Comment:
lol i was debugging this exact line of code since user_msg and tools were swapped

Reply:
Lol yes, caught it later on, whoops